homeassistant/components/telegram_bot/telegram_poll.py

```diff
@@ -0,0 +1,120 @@
+"""Telegram bot polling implementation."""
+
+import asyncio
+from asyncio.futures import CancelledError
+import logging
+
+
+from homeassistant.components.telegram_bot import CONF_ALLOWED_CHAT_IDS, \
+    process_message
+from homeassistant.const import EVENT_HOMEASSISTANT_START, \
+    EVENT_HOMEASSISTANT_STOP, CONF_API_KEY
+from homeassistant.core import callback
+from homeassistant.helpers.aiohttp_client import async_get_clientsession
+from aiohttp.errors import ClientError, ClientDisconnectedError
+_LOGGER = logging.getLogger(__name__)
+
+REQUIREMENTS = ['python-telegram-bot==5.3.0']
+
+
+@asyncio.coroutine
+def async_setup_platform(hass, config, async_add_devices, discovery_info=None):
+    """Setup the polling platform."""
+    import telegram
+    bot = telegram.Bot(config[CONF_API_KEY])
+    allowed_chat_ids = config[CONF_ALLOWED_CHAT_IDS]
+    pol = TelegramPoll(bot, hass, allowed_chat_ids)
+
+    @callback
+    def _start_bot(_event):
+        """Start the bot."""
+        pol.start_polling()
+
+    @callback
+    def _stop_bot(_event):
+        """Stop the bot."""
+        pol.stop_polling()
+
+    hass.bus.async_listen_once(
+        EVENT_HOMEASSISTANT_START,
+        _start_bot
+    )
+    hass.bus.async_listen_once(
+        EVENT_HOMEASSISTANT_STOP,
+        _stop_bot
+    )
+    async_add_devices(pol)
+
+
+class TelegramPoll:
+    """asyncio telegram incoming message handler."""
+
+    def __init__(self, bot, hass, allowed_chat_ids):
+        """Initialize the polling instance."""
+        self.allowed_chat_ids = allowed_chat_ids
+        self.update_id = 0
+        self.hass = hass
+        self.websession = async_get_clientsession(hass)
+        self.update_url = '{0}/getUpdates'.format(bot.base_url)
+        self.polling_task = None  # The actuall polling task.
+
+    def start_polling(self):
+        """Start the polling task."""
+        self.polling_task = self.hass.async_add_job(self.check_incoming())
+
+    def stop_polling(self):
+        """Stop the polling task."""
+        self.polling_task.cancel()
+
+    @asyncio.coroutine
+    def get_updates(self, offset, timeout):
+        """Bypass the default longpolling method to enable asyncio."""
+        resp = None
+        _json = []  # The actual value to be returned.
+        data = {'timeout': timeout}
+        if offset:
+            data['offset'] = offset
+        try:
+            resp = yield from self.websession.post(
```

Comment:
Thought about that, but I figured the timeout in the request body would take care of that. Adding an extra one looked just strange. But I can add one with a timeout slightly longer than the one in the request body...